app/assets/javascripts/admin/addon/templates/admin-badges/show.hbs

```diff
@@ -88,9 +88,9 @@
       </div>
 
       {{#if this.hasQuery}}
-        <a href {{action "preview" this.buffered "false"}}>{{i18n "admin.badges.preview.link_text"}}</a>
+        <a href {{on "click" (prevent-default (fn this.preview this.buffered "false"))}}>{{i18n "admin.badges.preview.link_text"}}</a>
```

Comment:
@davidtaylorhq How do you feel about using `(prevent-default)` helper?

If possible I think I'd prefer plain old `event.preventDefault()` inside each action? (which we already do in some of them)

Nesting `{{on}}`, `(prevent-default)`, `(fn)` and (sometimes) `(mut)` can get a bit unreadable.

Reply:
> Nesting {{on}}, (prevent-default), (fn) and (sometimes) (mut) can get a bit unreadable.

I can definitely see this point. To take one particularly rough example:

```diff
-    <a href {{action (mut this.pollType) "regular"}} class="poll-type-value {{if this.isRegular "active"}}">
+    <a href {{on "click" (prevent-default (action (mut this.pollType) "regular"))}} class="poll-type-value {{if this.isRegular "active"}}">
```

I used `prevent-default` to make the refactor away from the `action` modifier as mechanical as possible on a first pass, because it allowed changes to remain in templates for the most part. If you'd like me to reduce / eliminate usage of `prevent-default` to de-clutter the templates, I could do so via `event.preventDefault()` in action methods as part of this PR. It will require some more subjective refactoring and the introduction of a number of new methods, but I'd be glad to take a shot at this if you like.

Reply:
My main concern is that introducing `ember-event-helpers` and adding many uses in core will encourage theme/plugin developers to start using it. And if that happens, we'll be stuck supporting it for a long time, even after we clean up core.

That said, I realise that there are 103 uses of `prevent-default` in this PR, so it'll be no small feat to refactor them all into methods.

Maybe take a couple of hours and see how far you get @dgeb? If it's looking feasible then we can proceed. If not, we can rethink.

Reply:
@CvX @davidtaylorhq I spent some time today on [this commit](https://github.com/discourse/discourse/pull/18333/commits/dc45c297a18b0f80687b1f1c4ee8e547c51f10c9) to refactor away usages of `prevent-default` in favor of action handlers that invoke `event.preventDefault()`. I would say that I'm over 2/3 of the way done with this refactor and should be able to easily wrap it soon.

In general, the refactor really cleans up templates, with changes such as:

```diff
-              <a {{on "click" (prevent-default (action "close"))}} href={{user.path}} class="card-tiny-avatar">{{bound-avatar user "tiny"}}</a>
+               <a {{on "click" this.close}} href={{user.path}} class="card-tiny-avatar">{{bound-avatar user "tiny"}}</a>
```

However, one of the interesting implications here is that, because action handlers must accept `event` in order to invoke `event.preventDefault()`, then other arguments must be passed indirectly via state on `event.target` (i.e. the target element). For example, in tag-info.hbs note the addition of `data-tag-id`:

```diff
-                <a href {{on "click" (prevent-default (action "deleteSynonym" tag))}} class="delete-synonym">
+                 <a href {{on "click" this.deleteSynonym}} data-tag-id={{tag.id}} class="delete-synonym">
```

This `data-tag-id` attribute is now extracted in the `deleteSynonym` action handler:

```js
   @action
   deleteSynonym(event) {
     event.preventDefault();
     const tagId = event.target.getAttribute("data-tag-id");
     const tag = this.tagInfo.synonyms.findBy("id", tagId);
     // ...
   },
```

This is a fairly common pattern but it is a pretty big delta from passing args directly to the action. If we want to continue passing args directly instead of using data attributes, I can think of a couple different approaches to consider:

* Continue using the `prevent-default` helper sparingly where an action handler requires args other than `event`.
* Move from `a` to `button` tags for actions to be more conventional and to avoid the need for any `preventDefault` handling at all. This would allow event handlers such as: 
`<button {{on "click" (fn this.deleteSynonym tag)}}>Delete</button>`

I thought I'd pause for the day to give you a chance to review this work and think over these options. Let me know if you like the current direction of completely removing `prevent-default` or whether you want to take a slightly different approach.

Reply:
Thanks for the suggestion @CvX and @davidtaylorhq – your "Plan C" sounds like the right approach here :)

I think I've had a blind spot here because arg concatenation by `fn` in general seems like a potential footgun that can lead to surprises due to function arity. Although it's not a pattern I normally rely on, in this case it's clearly useful and I think that it will have the least impact on the core codebase as well as theme/plugin compatibility. 

I'll finish this refactor soon and try to minimize risk by always treating `event` as optional in handler args (e.g. `event?.preventDefault()`).

Reply:
Marking this convo as resolved now that I've completely refactored this PR away from using `ember-event-helpers` and `prevent-default` in particular.